packages/viewer/src/scenes/operations.ts

```diff
@@ -0,0 +1,241 @@
+  }
+
+  public and(): SelectorBuilder<AndSelector> {
+    return new AndSelectorBuilder(this);
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.query = { type: 'supplied-id', value: suppliedId };
+    return this;
+  }
+
+  public withItemId(itemId: string): this {
+    this.query = { type: 'item-id', value: itemId };
+    return this;
+  }
+}
+
+/**
+ * A selector builder to perform boolean `or` operations.
+ */
+export class OrSelectorBuilder implements SelectorBuilder<OrSelector> {
+  private builders?: SelectorBuilder<ItemSelector>[] = [];
+
+  private parent: SelectorBuilder<ItemSelector>;
+  public constructor(parent: SelectorBuilder<ItemSelector>) {
+    this.builders.push(parent);
+    this.parent = parent;
+  }
+
+  public or(): SelectorBuilder<OrSelector> {
+    return this;
+  }
+
+  public and(): SelectorBuilder<AndSelector> {
+    return new AndSelectorBuilder(this.parent);
+  }
+
+  public build(): OrSelector {
+    return { type: 'or', selectors: this.builders.map(q => q.build()) };
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withSuppliedId(suppliedId));
+    return this;
+  }
+
+  public withItemId(partId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withItemId(partId));
+    return this;
+  }
+}
+
+/**
+ * A selector builder to perform boolean `and` operations.
+ */
+export class AndSelectorBuilder implements SelectorBuilder<AndSelector> {
+  private builders?: SelectorBuilder<ItemSelector>[] = [];
+  private parent: SelectorBuilder<ItemSelector>;
+  public constructor(parent: SelectorBuilder<ItemSelector>) {
+    this.builders.push(parent);
+    this.parent = parent;
+  }
+
+  public or(): SelectorBuilder<OrSelector> {
+    return new OrSelectorBuilder(this.parent);
+  }
+
+  public and(): SelectorBuilder<AndSelector> {
+    return this;
+  }
+
+  public build(): AndSelector {
+    return { type: 'and', selectors: this.builders.map(q => q.build()) };
+  }
+
+  public withSuppliedId(suppliedId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withSuppliedId(suppliedId));
+    return this;
+  }
+
+  public withItemId(partId: string): this {
+    this.builders.push(new ItemSelectorBuilder().withItemId(partId));
+    return this;
+  }
+}
+
+/**
+ * An `ItemSelector` describes a type for performing a selection against some
+ * attribute of a scene item.
+ */
+export type ItemSelector = SuppliedIdSelector | ItemIdSelector;
+
+interface OrSelector {
+  type: 'or';
+  selectors: ItemSelector[];
+}
+
+interface AndSelector {
+  type: 'and';
+  selectors: ItemSelector[];
+}
+
+export interface OperationDefinition {
+  operation: ItemOperation | SceneOperation;
+}
+
+export interface SelectorBuilder<T> {
+  /**
+   * Returns the built selector.
+   */
+  build(): T;
+
+  /**
+   * Returns a conditional builder to perform an `or` operation.
+   */
+  or(): SelectorBuilder<OrSelector>;
```

Comment:
I'd recommend splitting the operations (`or`, `and`, etc) and the `build` method into separate interfaces. Maybe `SceneQuery`? Then `SelectorBuilder` could extend from this new interface. This way the types won't allow you to call `.build()` inside a where clause.

Reply:
I have split these out now, but am having a bit of trouble converting my `Selector` to a `SelectorBuilder`. Was wondering if you could provide any suggestions or insight into this problem?